tests/jax/components/training/model_updating_test.py

```diff
@@ -0,0 +1,394 @@
+
+    assert mock_trainer.store.optimizer.test == "Done"
+
+    for net_key in mock_trainer.store.networks["networks"].keys():
+        assert mock_trainer.store.opt_states[net_key] == list(
+            mock_trainer.store.networks["networks"][net_key].params
+        )
+
+    assert mock_trainer.store.minibatch_update_fn
+
+
+def test_minibatch_update_fn(
+    mock_state_and_trainer: Tuple[Dict[str, Any], MockTrainer]
+) -> None:
+    """Test on_minibatch_update_fn
+
+    Args:
+        mock_state_and_trainer: tuple include fake state and mock trainer
+    """
+    state = mock_state_and_trainer[0]
+    mock_trainer = mock_state_and_trainer[1]
+    carry = [state["params"], state["opt_states"]]
+    (
+        new_params,
+        new_opt_states,
+    ), metrics = mock_trainer.store.minibatch_update_fn(
+        carry=carry, minibatch=state["batch"]
+    )
+
+    assert list(new_params.keys()) == [
+        "network_agent_0",
+        "network_agent_1",
+        "network_agent_2",
+    ]
+    assert list(new_params["network_agent_0"]) == [5.0, 5.0, 5.0]
+    assert list(new_params["network_agent_1"]) == [6.0, 6.0, 6.0]
+    assert list(new_params["network_agent_2"]) == [7.0, 7.0, 7.0]
+
+    for net_key in new_opt_states.keys():
+        assert new_opt_states[net_key] == "opt_states_after_update"
+
+    assert list(metrics.keys()) == ["agent_0", "agent_1", "agent_2"]
+    for agent in metrics.keys():
+        assert list(metrics[agent].keys()) == ["norm_grad", "norm_updates"]
+        assert metrics[agent]["norm_grad"] == optax.global_norm(
+            jnp.array([5.0, 5.0, 5.0])
+        )
+        assert metrics[agent]["norm_updates"] == optax.global_norm(
+            jnp.array([5.0, 5.0, 5.0])
+        )
+
+
+def test_on_training_utility_fns_epoch(mock_trainer: MockTrainer) -> None:
+    """Test on_training_utility_fns from MAPGEpochUpdate
+
+    Args:
+        mock_trainer: trainer
+    """
+    mini_epoch_update = MAPGEpochUpdate()
+    mini_epoch_update.on_training_utility_fns(trainer=mock_trainer)
+
+    assert mock_trainer.store.num_epochs == mini_epoch_update.config.num_epochs
+    assert (
+        mock_trainer.store.num_minibatches == mini_epoch_update.config.num_minibatches
+    )
+    assert mock_trainer.store.epoch_update_fn
+
+
+def test_epoch_update_fn(
+    mock_state_and_trainer: Tuple[Dict[str, Any], MockTrainer],
+    mock_minibatch_update_fn: Callable,
+) -> None:
+    """Test epoch_update_fn function
+
+    Args:
+        mock_state_and_trainer: tuple include fake state and mock trainer
+        mock_minibatch_update_fn
+    """
+    mock_trainer = mock_state_and_trainer[1]
+    mini_epoch_update = MAPGEpochUpdate()
+    mini_epoch_update.on_training_utility_fns(trainer=mock_trainer)
+    mock_trainer.store.minibatch_update_fn = mock_minibatch_update_fn
+
+    state = mock_state_and_trainer[0]
+    # update params and opt_state to function in jax.lax.scan function
+    state["params"] = jnp.array([0, 0, 0])
```

Comment:
I think this will make the code longer since I am going to build another mock because I tried to make only one mock to be adapted to all the functions testings, what do you think if it is a really important to change I will give it a try?

Reply:
I think if changing this affects another test then either leave it as is or try get correct asserts for the state defined in the mock in this test. But leaving is probably fine.